### ISSUE TITLE
np.resize crashes when the product of the array dimensions overflows an int32

### ISSUE THREAD

**Issue by adamreeve** [numpy/numpy#17080]
This looks to be related to #12664 which was closed because numpy.core.umath.multiply is expected to overflow, but I would expect that resize should work without the data shape array having to be int64.

### Reproducing code example:

```python
import numpy as np

array_length = 2 ** 31 + 4
data_shape = (array_length // 4, 2, 2)
assert np.product(data_shape, dtype=np.dtype('int64')) == array_length
data = np.zeros(array_length, dtype=np.dtype('int8'))

resized = np.resize(data, data_shape)  # exception thrown here
```

If I change the data shape so that it is an array with dtype of int64 (eg. add `data_shape = np.array(data_shape, dtype=np.int64)` before the call to resize) then this does work as expected.

### Error message:

```
Traceback (most recent call last):
  File ".\repro.py", line 12, in <module>
    resized = np.resize(data, data_shape)
  File "<__array_function__ internals>", line 5, in resize
  File "C:\Python38\lib\site-packages\numpy\core\fromnumeric.py", line 1421, in resize
    return reshape(a, new_shape)
  File "<__array_function__ internals>", line 5, in reshape
  File "C:\Python38\lib\site-packages\numpy\core\fromnumeric.py", line 299, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File "C:\Python38\lib\site-packages\numpy\core\fromnumeric.py", line 58, in _wrapfunc
    return bound(*args, **kwds)
ValueError: cannot reshape array of size 8 into shape (536870913,2,2)
```

### Numpy/Python version information:

1.19.1 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 23:03:10) [MSC v.1916 64 bit (AMD64)]

I can reproduce this on Windows but not on Linux.

**Comment by eric-wieser** [comment]
In this particular case you want `reshape` not `resize`

**Comment by eric-wieser** [comment]
Although I think `tensor_prod` has a similar bug.

**Comment by adamreeve** [comment]
> In this particular case you want reshape not resize

Yes after encountering this bug I changed the code so it didn't use `resize` but thought it would be useful to report the issue.

Is there an easy way to verify this is fixed? Looks like I'd need to build numpy from source.

**Comment by mattip** [comment]
> Is there an easy way to verify this? 

Maybe we should make the weekly builds at https://anaconda.org/scipy-wheels-nightly/numpy more discoverable. You should be able to do `pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy`

**Comment by adamreeve** [comment]
Thanks, I can confirm it's fixed with numpy 1.20.0.dev0+eec0aa2 from that index so will close this.
